File: java/advanced/FenwickTree.java
package advanced;

import java.util.Arrays;

/**
 * Fenwick Tree (Binary Indexed Tree) for dynamic prefix-sum queries.
 *
 * <p>The Fenwick Tree stores partial sums in an array {@code bit[]} indexed from 1.
 * The "responsible range" of node {@code i} is determined by the lowest set bit of {@code i}:
 * {@code LSB(i) = i & (-i)}.  Node {@code i} accumulates the sum of elements
 * {@code [i - LSB(i) + 1, i]}.
 *
 * <p>Time complexities:
 * <ul>
 *   <li>update     – O(log n)</li>
 *   <li>prefixSum  – O(log n)</li>
 *   <li>rangeQuery – O(log n)</li>
 *   <li>build      – O(n log n) naive, O(n) with the optimised constructor</li>
 * </ul>
 *
 * <p>Space complexity: O(n).
 */
public class FenwickTree {

    // -------------------------------------------------------------------------
    // Fields
    // -------------------------------------------------------------------------

    private final int[] bit; // 1-indexed Fenwick array
    private final int   n;   // size of the logical array

    // -------------------------------------------------------------------------
    // Constructors
    // -------------------------------------------------------------------------

    /**
     * Creates a Fenwick Tree of size {@code n} with all values set to zero.
     *
     * <p>Time: O(n) | Space: O(n).
     *
     * @param n the number of elements (1-indexed up to n)
     */
    public FenwickTree(int n) {
        if (n <= 0) throw new IllegalArgumentException("Size must be positive");
        this.n   = n;
        this.bit = new int[n + 1];
    }

    /**
     * Creates a Fenwick Tree initialised from {@code arr} (0-indexed).
     *
     * <p>Uses an O(n) construction: copy values into {@code bit} and propagate
     * each node's value up to its parent exactly once.
     *
     * <p>Time: O(n) | Space: O(n).
     *
     * @param arr source array (0-indexed)
     */
    public FenwickTree(int[] arr) {
        this(arr.length);
        // O(n) build: set bit[i] = arr[i-1], then propagate
        for (int i = 1; i <= n; i++) {
            bit[i] += arr[i - 1];
            int parent = i + (i & -i);
            if (parent <= n) bit[parent] += bit[i];
        }
    }

    // -------------------------------------------------------------------------
    // update
    // -------------------------------------------------------------------------

    /**
     * Adds {@code delta} to the element at 1-indexed position {@code i}.
     * To set a value: call {@code update(i, newVal - oldVal)}.
     *
     * <p>Time: O(log n) | Space: O(1).
     *
     * @param i     1-indexed position (1 &le; i &le; n)
     * @param delta the amount to add
     */
    public void update(int i, int delta) {
        checkIndex(i);
        for (; i <= n; i += i & (-i)) {
            bit[i] += delta;
        }
    }

    // -------------------------------------------------------------------------
    // prefixSum
    // -------------------------------------------------------------------------

    /**
     * Returns the prefix sum of elements {@code [1, i]} (1-indexed).
     *
     * <p>Time: O(log n) | Space: O(1).
     *
     * @param i 1-indexed upper bound (1 &le; i &le; n)
     * @return sum of elements from index 1 to i
     */
    public int prefixSum(int i) {
        checkIndex(i);
        int sum = 0;
        for (; i > 0; i -= i & (-i)) {
            sum += bit[i];
        }
        return sum;
    }

    // -------------------------------------------------------------------------
    // rangeQuery
    // -------------------------------------------------------------------------

    /**
     * Returns the sum of elements in the range {@code [l, r]} (1-indexed, inclusive).
     *
     * <p>Time: O(log n) | Space: O(1).
     *
     * @param l left bound (1-indexed, inclusive)
     * @param r right bound (1-indexed, inclusive)
     * @return sum of elements from index l to r
     */
    public int rangeQuery(int l, int r) {
        if (l < 1 || r > n || l > r)
            throw new IndexOutOfBoundsException("Invalid range [" + l + ", " + r + "]");
        return prefixSum(r) - (l > 1 ? prefixSum(l - 1) : 0);
    }

    // -------------------------------------------------------------------------
    // Helpers
    // -------------------------------------------------------------------------

    private void checkIndex(int i) {
        if (i < 1 || i > n)
            throw new IndexOutOfBoundsException("Index " + i + " out of range [1, " + n + "]");
    }

    /** Returns the size of the Fenwick Tree. */
    public int size() { return n; }

    // -------------------------------------------------------------------------
    // toString
    // -------------------------------------------------------------------------

    /**
     * Returns a string showing the internal {@code bit[]} array and the logical
     * values it represents, with each node's responsible range.
     *
     * <p>Time: O(n log n) | Space: O(n).
     *
     * @return multi-line Fenwick Tree representation
     */
    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("FenwickTree (n=").append(n).append("):\n");

        // Reconstruct logical array via prefix sums
        int[] logical = new int[n + 1];
        for (int i = 1; i <= n; i++) {
            logical[i] = rangeQuery(i, i);
        }
        sb.append("  Logical array (1-indexed): ");
        sb.append("[_, ");
        for (int i = 1; i <= n; i++) {
            sb.append(logical[i]);
            if (i < n) sb.append(", ");
        }
        sb.append("]\n");

        sb.append("  bit[] (1-indexed)         : ");
        sb.append("[_, ");
        for (int i = 1; i <= n; i++) {
            sb.append(bit[i]);
            if (i < n) sb.append(", ");
        }
        sb.append("]\n");

        // Node details
        sb.append("  Node details:\n");
        for (int i = 1; i <= n; i++) {
            int lsb = i & (-i);
            int rangeStart = i - lsb + 1;
            sb.append("    bit[").append(i).append("] = ").append(bit[i])
              .append("  (covers [").append(rangeStart).append(", ").append(i).append("])\n");
        }
        return sb.toString();
    }

    // -------------------------------------------------------------------------
    // Main – demo
    // -------------------------------------------------------------------------

    public static void main(String[] args) {
        System.out.println("=== Fenwick Tree (BIT) Demo ===\n");

        // Build from array (1-indexed usage: positions 1..8)
        int[] arr = {1, 3, 5, 7, 9, 11, 13, 15}; // 1-indexed: pos 1..8
        FenwickTree ft = new FenwickTree(arr);
        System.out.println(ft);

        // Prefix sums
        System.out.println("prefixSum(4) = " + ft.prefixSum(4) + "  (expected 16: 1+3+5+7)");
        System.out.println("prefixSum(8) = " + ft.prefixSum(8) + "  (expected 64: sum all)");
        System.out.println("prefixSum(1) = " + ft.prefixSum(1) + "  (expected 1)");

        // Range queries
        System.out.println("\nrangeQuery(2, 5) = " + ft.rangeQuery(2, 5) + "  (expected 24: 3+5+7+9)");
        System.out.println("rangeQuery(1, 8) = " + ft.rangeQuery(1, 8) + "  (expected 64)");
        System.out.println("rangeQuery(3, 3) = " + ft.rangeQuery(3, 3) + "  (expected 5)");

        // Update
        System.out.println("\nupdate(3, 4)  -- add 4 to index 3 (5 -> 9)");
        ft.update(3, 4);
        System.out.println("prefixSum(4) = " + ft.prefixSum(4) + "  (expected 20: 1+3+9+7)");
        System.out.println("rangeQuery(2, 5) = " + ft.rangeQuery(2, 5) + "  (expected 28: 3+9+7+9)");

        System.out.println("\nAfter update:");
        System.out.println(ft);

        // Build from size and individual updates
        System.out.println("Fresh FenwickTree(5), updating positions 1..5:");
        FenwickTree ft2 = new FenwickTree(5);
        for (int i = 1; i <= 5; i++) ft2.update(i, i * 2);
        System.out.println(ft2);
        System.out.println("prefixSum(5) = " + ft2.prefixSum(5) + "  (expected 30: 2+4+6+8+10)");
    }
}

File: python/advanced/fenwick_tree.py
"""
Fenwick Tree (Binary Indexed Tree — BIT)
==========================================
A clever array-based structure for prefix-sum queries and point updates.
Each index i is responsible for a range of elements determined by its lowest
set bit (LSB): the range length equals LSB(i) = i & (-i).

Navigation:
    - To move to the parent (for update): i += i & (-i)   # add LSB
    - To move to the responsible prefix  (for query): i -= i & (-i)   # remove LSB

This file uses **1-based indexing** (standard BIT convention).
External API uses 0-based indexes for Python friendliness.

Complexities:
    - update(i, delta):   O(log n)
    - prefix_sum(i):      O(log n)
    - range_query(l, r):  O(log n)   [two prefix_sum calls]
    - build from list:    O(n log n)  [n updates], or O(n) with the trick below
    - Space:              O(n)
"""

from __future__ import annotations
from typing import List


class FenwickTree:
    """
    Binary Indexed Tree for prefix-sum queries and point updates.

    All external indices are 0-based.
    Internally uses 1-based indices for correct LSB arithmetic.
    """

    def __init__(self, n: int) -> None:
        """
        Create an empty BIT of size *n*.

        Time:  O(n)
        """
        self._n = n
        self._tree: List[int] = [0] * (n + 1)  # 1-indexed; index 0 unused

    # ------------------------------------------------------------------
    # Core operations
    # ------------------------------------------------------------------

    def update(self, i: int, delta: int) -> None:
        """
        Add *delta* to position *i* (0-based).

        Internally traverses upward by adding the LSB repeatedly:
            i, i + (i & -i), ...

        BIT Bit Trick:
            i & (-i) isolates the lowest set bit.
            Adding it moves i to the next node that is responsible for
            a range that includes the original i.

        Time:  O(log n)

        Example for n=8, i=3 (1-based=4, binary=0100):
            Step 1: pos=4  (0100), update tree[4]
            Step 2: pos=8  (1000), update tree[8]   (+4 = add LSB 0100)
            (8 > n=8? no, so done)
        """
        pos = i + 1  # convert to 1-based
        while pos <= self._n:
            self._tree[pos] += delta
            # Move to next responsible ancestor.
            pos += pos & (-pos)  # add lowest set bit

    def prefix_sum(self, i: int) -> int:
        """
        Return the sum of arr[0..i] (inclusive, 0-based).

        Traverses downward by removing the LSB:
            i, i - (i & -i), ...

        BIT Bit Trick:
            i & (-i) isolates the lowest set bit.
            Removing it moves i to the parent in the BIT sense —
            the node whose range ends just before this one begins.

        Time:  O(log n)

        Example for i=6 (1-based=7, binary=0111):
            Step 1: pos=7  (0111), accumulate tree[7]
            Step 2: pos=6  (0110), accumulate tree[6]  (-1 = remove LSB 0001)
            Step 3: pos=4  (0100), accumulate tree[4]  (-2 = remove LSB 0010)
            Step 4: pos=0, stop.
        """
        total = 0
        pos = i + 1  # convert to 1-based
        while pos > 0:
            total += self._tree[pos]
            # Move to parent — strip lowest set bit.
            pos -= pos & (-pos)  # remove lowest set bit
        return total

    def range_query(self, l: int, r: int) -> int:
        """
        Return the sum of arr[l..r] (inclusive, 0-based).

        Uses the identity: sum(l, r) = prefix_sum(r) - prefix_sum(l-1).

        Time:  O(log n)
        """
        if l > r:
            raise ValueError(f"l={l} must be <= r={r}")
        if l == 0:
            return self.prefix_sum(r)
        return self.prefix_sum(r) - self.prefix_sum(l - 1)

    # ------------------------------------------------------------------
    # Build from existing list
    # ------------------------------------------------------------------

    @classmethod
    def build(cls, data: List[int]) -> "FenwickTree":
        """
        Build a Fenwick Tree from a list in O(n) time.

        Faster O(n) trick: directly compute tree[i] = partial sum
        that each index is responsible for, rather than calling update n times.

        Time:  O(n)
        Space: O(n)
        """
        n = len(data)
        ft = cls(n)
        # Copy data into 1-indexed positions.
        for i in range(1, n + 1):
            ft._tree[i] = data[i - 1]
        # Propagate each position's value to its parent.
        for i in range(1, n + 1):
            parent = i + (i & (-i))   # add lowest set bit → parent
            if parent <= n:
                ft._tree[parent] += ft._tree[i]
        return ft

    # ------------------------------------------------------------------
    # Point-value accessor
    # ------------------------------------------------------------------

    def point_value(self, i: int) -> int:
        """
        Return the current value at index *i* in O(log n) via two prefix queries.

        Time:  O(log n)
        """
        if i == 0:
            return self.prefix_sum(0)
        return self.prefix_sum(i) - self.prefix_sum(i - 1)

    # ------------------------------------------------------------------
    # String representation
    # ------------------------------------------------------------------

    def __repr__(self) -> str:
        return f"FenwickTree(n={self._n}, internal={self._tree[1:]})"


# ----------------------------------------------------------------------
# Application: Kruskal's MST helper (no actual graph — conceptual demo)
# ----------------------------------------------------------------------

def count_inversions(arr: List[int]) -> int:
    """
    Count inversions in *arr* using a Fenwick Tree.

    An inversion is a pair (i, j) with i < j and arr[i] > arr[j].
    Useful in merge-sort-based counting but BIT gives a cleaner solution.

    Time:  O(n log n)
    Space: O(max_val)
    """
    if not arr:
        return 0
    max_val = max(arr)
    ft = FenwickTree(max_val + 1)
    inversions = 0
    for val in arr:
        # Count elements already inserted that are greater than val.
        inversions += ft.range_query(val + 1, max_val) if val < max_val else 0
        ft.update(val, 1)
    return inversions


# ----------------------------------------------------------------------
# Demo
# ----------------------------------------------------------------------

if __name__ == "__main__":
    print("=== Fenwick Tree (BIT) ===")
    data = [1, 7, 3, 0, 7, 8, 3, 2, 6, 2]
    print("Data:", data)

    ft = FenwickTree.build(data)
    print()
    print("prefix_sum(0)  :", ft.prefix_sum(0))   # 1
    print("prefix_sum(4)  :", ft.prefix_sum(4))   # 18
    print("prefix_sum(9)  :", ft.prefix_sum(9))   # 39
    print()
    print("range_query(2,5):", ft.range_query(2, 5))   # 3+0+7+8 = 18
    print("range_query(0,9):", ft.range_query(0, 9))   # 39

    print()
    print("Updating index 3: value 0 → 5 (delta=+5)")
    ft.update(3, 5)
    print("range_query(2,5):", ft.range_query(2, 5))   # 3+5+7+8 = 23
    print("prefix_sum(9)   :", ft.prefix_sum(9))        # 44

    print()
    print("=== Inversion Count ===")
    inv_arr = [8, 4, 2, 1]
    print(f"Inversions in {inv_arr}:", count_inversions(inv_arr))  # 6

    inv_arr2 = [3, 1, 2]
    print(f"Inversions in {inv_arr2}:", count_inversions(inv_arr2))  # 2

    print()
    print("Internal BIT array:", ft)
